spike_main/spike.cc

```diff
@@ -118,28 +118,42 @@ static void read_file_bytes(const char *filename,size_t fileoff,
 
 bool sort_mem_region(const mem_cfg_t &a, const mem_cfg_t &b)
 {
-  if (a.base == b.base)
-    return (a.size < b.size);
+  if (a.get_base() == b.get_base())
+    return (a.get_size() < b.get_size());
   else
-    return (a.base < b.base);
+    return (a.get_base() < b.get_base());
 }
 
 static bool check_mem_overlap(const mem_cfg_t& L, const mem_cfg_t& R)
 {
-  const reg_t L_end   = L.base + L.size - 1;
-  const reg_t R_end   = R.base + R.size - 1;
+  return std::max(L.get_base(), R.get_base()) <= std::min(L.get_inclusive_end(), R.get_inclusive_end());
+}
+
+// there is a weird corner-case preventing two memory regions to be merged -
+// this can happen when the resulting size of a region is 2^64.
+// such regions are not representable with mem_cfg_t class
+static bool check_if_merge_covers_64bit_space(const mem_cfg_t& L,
+                                              const mem_cfg_t& R)
+{
+  if (!check_mem_overlap(L, R))
+    return false;
 
-  return std::max(L.base, R.base) <= std::min(L_end, R_end);
+  auto start = std::min(L.get_base(), R.get_base());
+  auto end = std::max(L.get_inclusive_end(), R.get_inclusive_end());
+
+  // if [base == 0] and [inclusive_end == (2 ^ 64 - 1)] - we can't merge such
+  // regions, since the mem_cfg_t does not support the resulting size
```

Comment:
I think this comment better belongs down on line 181, along with an explanation of the workaround, such as: "So create two smaller memories that total 2^64 bytes."